[CSL-2433] Wrap actions with MaybeT to enable gaurds to exit gracefully

akegalj opened the pull request:
This PR fixes behaviour of `guard` in integration tests. Previously `guard` was using implementation of `Alternative IO` which would fail : 

```haskell
-- | @since 4.9.0.0
instance Alternative IO where
    empty = failIO "mzero"
    (<|>) = mplusIO
```

and now integration tests are wrapped with `MaybeT` which returns `Nothing` if `guard` fails and thus implements correct behaviour for our use case - which is: "Ignore actions which don't satisfy `guard`". We can simply achieve this with `catMaybes` or similar.

Issue: https://iohk.myjetbrains.com/youtrack/issue/CSL-2433

_Note that PR will be merged to `feature/csl2398-integration-ci` branch and the goal of this PR is to have isolated place to review yt issue and discuss things._

ksaric commented:
Maybe I'm missing something here, but this seems wrong if you plan to do what you commented here - https://iohk.myjetbrains.com/youtrack/issue/CSL-2433

For example, look at the definition of [guard](http://hackage.haskell.org/package/base-4.11.0.0/docs/Control-Monad.html#v:guard).

akegalj commented:
@ksaric 
It is not obvious to me that I am on a wrong track. Can you elaborate?

> In other words, if the guard condition is not met, the action will retry, and another one will be performed (using action distribution). There are several ways to achieve this.

we can retry actions that failed on guard, when a first pass of actions is over. Or we can just skipp them. Retry seems correct way to achieve this. 

So what I have implemented will return `Maybe action` if `guard` didn't fail or `Nothing` if `guard` failed. In the case `guard` failed, we will run other actions that will be in the queue and retry failed actions later. We can repeat this step until all actions in the actions queue succeed (suecceed in the sense no action will be `guard`ed). 

Let me try to demonstrate the idea :

We are running the action here https://github.com/input-output-hk/cardano-sl/blob/3a2cdef60e87857a76577af4a29042720cd0b8ec/wallet-new/integration/Main.hs#L55 : 
```haskell 
    _ <- runActionCheck
        walletClient
        walletState
        actionDistr
```

Instead of ^ we would have some fold of actions: 
```haskell
runAllActions :: WalletClient -> WalletState -> ActionDistribution -> m WalletState
runAllActions walletClient walletState actionDistr = do
  -- guardedActions are those which failed with `guard`
  (newWalletState, guardedActions) <- foldM (someMoreLogicHere . runActionCheck walletClient) (walletState, mempty) actionDistr
  case nonEmpty guardedActions of
    Just ga -> runAllActions newWalletClient walletState ga
    Nothing -> pure newWalletState
```

Note that ^ is just a pseudo code and custom loop might not be necessary. Also if some action is guarded infinitelly (for example - bug in the code) CI would probably timeout and exit with error. We can probably make it better to build a counter for max number of tries `runAllActions` will retry. When number of tries exceeds we exit with error.

ksaric commented:
@akegalj Currently, it doesn't contain those things you wrote about above and is not valid. Once you write the code and you think it works I can comment more.

parsonsmatt commented:
We have a stateful, non-deterministic test suite where each action has some precondition before we can meaningfully do it.

`Alternative` expresses this nicely -- "if the randomly selected action isn't possible, try this other one." `IO`'s `Alternative` instance is ``a <|> b = a `catch` \(_ :: IOException) -> b``, so we can use `IO` directly here. But, we do need to express `<|>`.

If we generate a list of actions to take (eg ~100), we can do `foldr (\action rest -> (action *> rest) <|> rest) (pure ()) :: Alternative m => [m a] -> m a`. This will try `action`, and if it succeeds, do `rest`. 

Since `IOException` is the only one caught by `<|>`, we will be able to still fail tests by throwing other exceptions.

akegalj commented:
@parsonsmatt 
good point. Then we can get rid of `MaybeT` that I have added.

Note that I have added some variants how we can run tests. I was not sure which could we prefer so I have added few variants. Maybe there will be useful to us - depending how we want to test exactly. They will differ only in some corner cases though.

parsonsmatt commented on scripts/test/wallet/integration.sh:
```diff
@@ -38,7 +38,7 @@ tmux new-session -s $sessionName -d "WALLET_DEBUG=1 scripts/launch/demo-with-wal
 # wait until cluster is fully up and running
 echo "Waiting 140 seconds until local cluster is ready..."
-sleep 140s
+sleep 40s
```
I dropped this down because it was taking *a long* time to run tests, and i was getting distracted with twitter and friends between runs! can revert

parsonsmatt commented on scripts/test/wallet/integration.sh:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env sh
+#!/usr/bin/env bash
```
was getting some weird failures so i made it bash

parsonsmatt commented on wallet-new/integration/Error.hs:
```diff
@@ -21,24 +20,22 @@ import           Cardano.Wallet.Client (ClientError)
     | InvalidTransactionState Transaction
     | InvalidTransactionFee EstimatedFees
-    | LocalTransactionsDiffer [Transaction]
+    | LocalTransactionsDiffer [Transaction] [Transaction]
```
All of the `Differ` exception types now contain the local and remote state.

parsonsmatt commented on wallet-new/integration/Functions.hs:
```diff
@@ -35,454 +37,446 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
-        & actionsNum +~ 1
+    -- increment successful actions
+    log "Success!"
+    actionsNum += 1
```
The `actionsNum += 1` modification was common to all actions, so I factored it out.

parsonsmatt commented on wallet-new/integration/Functions.hs:
```diff
@@ -35,454 +37,446 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
-    result      <- respToRes $ postWallet wc newWall
+runAction wc action = do
+    log $ "Running: " <> show action
+    case action of
```
I switched to an explicit `case` rather than a top level pattern match, because I wanted to log the action that was taken, and also reduce a tiny bit of the boilerplate.

parsonsmatt commented on wallet-new/integration/Functions.hs:
```diff
@@ -517,8 +522,8 @@ respToRes resp = do
-pickRandomElement :: (MonadIO m, HasCallStack) => [a] -> m a
-pickRandomElement [] = error "called with empty list"
+pickRandomElement :: (MonadIO m, Alternative m) => [a] -> m a
+pickRandomElement [] = empty
```
We were getting some `elements: called with empty list` errors. This will turn those showstoppers into test skips.

akegalj replied:
with this we can remove all explicit `guard`s

parsonsmatt commented on wallet-new/integration/Main.hs:
```diff
@@ -55,18 +51,23 @@ main = do
+
+initialWalletState :: WalletClient IO -> IO  WalletState
+initialWalletState wc = do
+    _wallets <- either throwM (pure . wrData) =<< getWallets wc
```
I was getting test failures where we'd do `GetWallets`, and there would *gasp* be some wallets already! so I initialize the state of the wallet with whatever the API returns. I'll likely need to get more of the stuff from the wallet initial state as well.

akegalj replied:
Hm that's strange

parsonsmatt commented on wallet-new/integration/Types.hs:
```diff
@@ -39,19 +34,14 @@ type WalletTestMode m =
+-- An 'Action' is paired with a 'Weight'. All of the 'Weight's are summed
+-- up, and the eventual probability is the weight divided by the total
+-- weight.
+newtype Weight = Weight { getWeight :: Int }
```
I switched from a `0-100` probability to a weight model. This drops all of the partiality involved and makes it simpler to work with.

parsonsmatt commented on wallet-new/src/Cardano/Wallet/API/V1/Errors.hs:
```diff
@@ -53,7 +53,7 @@ data WalletError =
     | MigrationFailed { weDescription :: !Text }
     | JSONValidationFailed { weValidationError :: !Text }
-    | UnkownError { weMsg :: !Text }
+    | UnknownError { weMsg :: !Text }
```
oh god, i typoed this :open_mouth:

akegalj commented on wallet-new/integration/Functions.hs:
```diff
@@ -35,454 +37,455 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
-    -- Just in case, let's check if it's still there.
-    result  <-  respToRes $ getWallets wc
+            -- The precondition is that we need to have wallets.
+            guard (not (null localWallets))
```
we don't even need to have this guard now, because `pickRandomElement` bellow in its current form will be the guard.

akegalj commented on wallet-new/integration/Functions.hs:
```diff
@@ -35,454 +37,455 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
-    pure $ ws
-        & actionsNum +~ 1
+            -- The precondition is that we need to have wallets.
+            guard (not (null localWallets))
```
I believe `pickRandomElement` will be the guard as well - so maybe this is not needed

akegalj commented on wallet-new/integration/Functions.hs:
```diff
@@ -35,454 +37,455 @@ import qualified Pos.Wallet.Web.ClientTypes.Types as V0
-    let oldWalletPass :: Maybe SpendingPassword
-        oldWalletPass = ws ^. walletsPass . at wallet
+            -- The precondition is that we need to have wallets.
+            guard (not (null localWallets))
```
probably not needed - `pickRandomElement` bellow will be the guard. 

It is used in many places so won't mention it again

akegalj commented:
@parsonsmatt 
> The script isn't cleanly restarting the server, so the tests aren't exactly idempotent.

I believe this is due too us runing deterministic tests before (and creating a wallet there). `NewWallet` should be fixed with https://iohk.myjetbrains.com/youtrack/issue/CSL-2432

akegalj commented on scripts/test/wallet/integration.sh:
```diff
@@ -31,27 +31,30 @@ rm -rf $tmpSecrets
-echo "Waiting 140 seconds until local cluster is ready..."
-sleep 140s
+echo "Waiting 40 seconds until local cluster is ready..."
+sleep 40s
```
40s doesn't work on my machine (80s does). It will be run by our CI which would probably handle 40 good enough

akegalj commented:
Merging to the base branch. Lets continue the work there
